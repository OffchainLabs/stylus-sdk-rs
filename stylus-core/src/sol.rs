use lazy_static::lazy_static;
use regex::Regex;

lazy_static! {
    static ref UINT_REGEX: Regex = Regex::new(r"^uint(\d+)$").unwrap();
    static ref INT_REGEX: Regex = Regex::new(r"^int(\d+)$").unwrap();
    static ref BYTES_REGEX: Regex = Regex::new(r"^bytes(\d+)$").unwrap();
}

pub fn is_sol_keyword(name: &str) -> bool {
    if let Some(caps) = UINT_REGEX.captures(name) {
        let bits: usize = caps[1].parse().unwrap();
        if bits.is_multiple_of(8) {
            return true;
        }
    }

    if let Some(caps) = INT_REGEX.captures(name) {
        let bits: usize = caps[1].parse().unwrap();
        if bits.is_multiple_of(8) {
            return true;
        }
    }

    if let Some(caps) = BYTES_REGEX.captures(name) {
        let bits: usize = caps[1].parse().unwrap();
        if bits <= 32 {
            return true;
        }
    }

    match name {
        // other types
        "address" | "bytes" | "bool" | "int" | "uint" => true,

        // other words
        "is" | "contract" | "interface" => true,

        // reserved keywords
        "after" | "alias" | "apply" | "auto" | "byte" | "case" | "copyof" | "default"
        | "define" | "final" | "implements" | "in" | "inline" | "let" | "macro" | "match"
        | "mutable" | "null" | "of" | "partial" | "promise" | "reference" | "relocatable"
        | "sealed" | "sizeof" | "static" | "supports" | "switch" | "typedef" | "typeof" | "var" => {
            true
        }
        _ => false,
    }
}
