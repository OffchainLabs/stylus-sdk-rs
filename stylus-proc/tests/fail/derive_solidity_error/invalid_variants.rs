// Copyright 2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

use alloy_sol_types::sol;

use stylus_proc::SolidityError;

sol! {
    error InsufficientBalance(address from, uint256 have, uint256 want);
    error InsufficientAllowance(address owner, address spender, uint256 have, uint256 want);
}

#[derive(SolidityError)]
enum MyError {
    Unit,
    Two(InsufficientBalance, InsufficientAllowance),
    Named { balance: InsufficientBalance },
}

fn main() {}
