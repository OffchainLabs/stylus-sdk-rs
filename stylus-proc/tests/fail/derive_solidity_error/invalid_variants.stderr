error: variant not a 1-tuple
  --> tests/fail/derive_solidity_error/invalid_variants.rs:15:5
   |
15 |     Unit,
   |     ^^^^

error: variant not a 1-tuple
  --> tests/fail/derive_solidity_error/invalid_variants.rs:16:8
   |
16 |     Two(InsufficientBalance, InsufficientAllowance),
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: variant not a 1-tuple
  --> tests/fail/derive_solidity_error/invalid_variants.rs:17:11
   |
17 |     Named { balance: InsufficientBalance },
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
